bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java

```diff
@@ -441,117 +335,187 @@ private void checkIfDirsOnSameDiskPartition(List<File> dirs) throws DiskPartitio
         }
     }
 
+    static Versioned<Cookie> readAndVerifyCookieFromRegistrationManager(Cookie masterCookie,
+                                                                        ServerConfiguration conf,
+                                                                        RegistrationManager rm)
+            throws BookieException {
+        // we need to loop through all possible bookie identifiers to ensure it is treated as a new environment
+        // just because of bad configuration
+        List<BookieSocketAddress> addresses = Lists.newArrayListWithExpectedSize(3);
+        try {
+            // ip address
+            addresses.add(getBookieAddress(
+                new ServerConfiguration(conf).setUseHostNameAsBookieID(false).setAdvertisedAddress(null)));
+            // host name
+            addresses.add(getBookieAddress(
+                new ServerConfiguration(conf).setUseHostNameAsBookieID(true).setAdvertisedAddress(null)));
+            // advertised address
+            if (null != conf.getAdvertisedAddress()) {
+                addresses.add(getBookieAddress(conf));
+            }
+        } catch (UnknownHostException e) {
+            throw new UnknownBookieIdException(e);
+        }
+        Versioned<Cookie> rmCookie = null;
+        for (BookieSocketAddress address : addresses) {
+            try {
+                rmCookie = Cookie.readFromRegistrationManager(rm, address);
```

Comment:
@sijie a bug. If there is a change in configuration, it is possible that we will write to a different znode than the one we read rmCookie from in this method. However, we will use the Version returned by this method, which will result in a BadVersionException.

Reply:
How's that possible after this change? Can you explain the sequence causing this bug?

Reply:
Say you have bookie, ip is 10.0.0.1, hostname is foobar.

First boot it comes up with setUseHostNameAsBookieID(false), writes its cookie to /ledgers/cookies/10.0.0.1:3181.
Second boot some disks are added, allow expansion is enabled and setUseHostNameAsBookieID is set to true. In this case, it reads a cookie from /ledgers/cookies/10.0.0.1:3181, updates the cookie for the new storage, writes to all the dirs and then tries to write the cookie to /ledgers/cookies/foobar:3181. However, this last write, it will use the version it read from  /ledgers/cookies/10.0.0.1:3181, which is wrong (it shouldn't even be a setData, it should be a create).

Reply:
btw, I think a good solution would be to just delete the old cookie before creating the new one, using a multi().

Reply:
If hostname is used and the IP cookie exists, the verification will already fail immediately at the first step. The second boot will never succeed. If you take a look at the test case changes, it covers your case described here.